cdap-security/src/main/java/co/cask/cdap/security/authorization/RemoteAuthorizationEnforcer.java

```diff
@@ -0,0 +1,193 @@
+                                         new DefaultHttpRequestConfig(false), "/v1/execute/");
+    this.cacheEnabled = cConf.getBoolean(Constants.Security.Authorization.CACHE_ENABLED);
+    int cacheTtlSecs = cConf.getInt(Constants.Security.Authorization.CACHE_TTL_SECS);
+    int cacheMaxEntries = cConf.getInt(Constants.Security.Authorization.CACHE_MAX_ENTRIES);
+
+    authPolicyCache = CacheBuilder.newBuilder()
+      .expireAfterWrite(cacheTtlSecs, TimeUnit.SECONDS)
+      .maximumSize(cacheMaxEntries)
+      .build(new CacheLoader<AuthorizationPrivilege, Boolean>() {
+        @SuppressWarnings("NullableProblems")
+        @Override
+        public Boolean load(AuthorizationPrivilege authorizationPrivilege) throws Exception {
+          LOG.trace("Cache miss for " + authorizationPrivilege);
+          return doEnforce(authorizationPrivilege);
+        }
+      });
+  }
+
+  @Override
+  public void enforce(EntityId entity, Principal principal, Action action) throws Exception {
+    if (!isSecurityAuthorizationEnabled()) {
+      return;
+    }
+    AuthorizationPrivilege authorizationPrivilege = new AuthorizationPrivilege(principal, entity, action);
+
+    boolean allowed = cacheEnabled ? authPolicyCache.get(authorizationPrivilege) : doEnforce(authorizationPrivilege);
+    if (!allowed) {
+      throw new UnauthorizedException(principal, action, entity);
+    }
+  }
+
+  private boolean doEnforce(AuthorizationPrivilege authorizationPrivilege) throws IOException {
+    HttpResponse response = executeRequest(authorizationPrivilege);
+    return HttpResponseStatus.OK.getCode() == response.getResponseCode();
+
+  }
+
+  private HttpResponse executeRequest(AuthorizationPrivilege authorizationPrivilege) throws IOException {
+    HttpRequest request = remoteClient.requestBuilder(HttpMethod.POST, "enforce")
+      .withBody(GSON.toJson(createBody(authorizationPrivilege.getEntityId(),
+                                       authorizationPrivilege.getPrincipal(),
+                                       authorizationPrivilege.getAction())))
+      .build();
+    return remoteClient.execute(request);
+  }
+
+  private static List<MethodArgument> createBody(Object... arguments) {
+    List<MethodArgument> methodArguments = new ArrayList<>();
+    for (Object arg : arguments) {
+      if (arg == null) {
+        methodArguments.add(null);
+      } else {
+        String type = arg.getClass().getName();
+        methodArguments.add(new MethodArgument(type, GSON.toJsonTree(arg)));
+      }
+    }
+    return methodArguments;
+  }
+
+  @VisibleForTesting
+  public Map<AuthorizationPrivilege, Boolean> cacheAsMap() {
+    return Collections.unmodifiableMap(authPolicyCache.asMap());
+  }
+
+  /**
+   * Key for caching Privileges on containers. This represents a specific privilege on which authorization can be
+   * enforced. The cache stores whether the enforce succeeded or failed.
+   */
+  public static class AuthorizationPrivilege {
```

Comment:
Not sure how to do that. We want to cache privileges for individual actions but we also want to let people grant and revoke on multiple actions.